### PR TITLE
Add block_list Option to set HTTP status for blocked flows/urls

### PR DESCRIPTION
**Pull request by ericbeland**
#### Description

This PR brings in a block_list Option that allows an HTTP status to be set for items on the blocklist. There are several cases where this can be handy, often around automated tests and performance measurement.

- Avoid overwhelming 3rd party websites and components that would usually be loaded by a page
- Avoid polluting analytics data with automated traffic
- Block ad networks or other traffic. If it is an ad network used by your own website, displaying ads to bots ruins your pricing/quality scores
- Quickly block image loads  block_list: [:~d  browserup.com & ~u png:200] so save bandwidth, etc.
- Quickly stub an ajax request to avoid a 404 that triggers a JS error callback

I modeled the PR after map_local.  I couldn't figure out how to get flake8 to run, though.

#### Checklist

 - [ x] I have updated tests where applicable.
 - [ x] I have added an entry to the CHANGELOG.

### CODE REVIEW

**Review by mhils — COMMENTED** [review]
Thanks for the PR! This generally looks useful. Some thoughts:

1. I don't think we need a separate `regex`, that's really what the filter can and should be used for. For map local it's different because there we `re.sub`, here we have a relatively constant response. Am I missing something?
2. You have a better perspective on this, but it might be generally useful to add a special case  for the non-standard [HTTP 444](https://httpstatuses.com/444) which kills the connection (`flow.kill()`) instead of returning a response. :)
3. We definitely need to add a documentation section at https://docs.mitmproxy.org/stable/overview-features/!
4. `tox -e flake8` :-)

**Comment by ericbeland** [comment]
1 - You're not missing anything, I hadn't considered that a filter with an & could get the same job done as a secondary regexp filter but also be more powerful.
2 - Good idea.
3 - I thought about it, but didn't know what the criteria was for when something merited a new section. Will do.
4 - Great!

**Comment by ericbeland** [comment]
In the latest version:
- added allow-only block type for restrictive blocking
- removed the regexp business
- added HTTP 444 status to the master list of response codes (even though it doesn't get sent, technically)
- added validation that checks the response code is real
- added special-case 444 status code and notes on it
- added overview for blocking
- fix flake issues

**Review by mhils — COMMENTED** [review]
Thanks for updating this! 🍰 

- What do we need to allow/block-all distinction for? Isn't that expressed as part of the filter again? For example, if I want to block all requests but example.com/foo, I set `!example.com/foo` as my filter.
- Let's move the documentation into its own section on the https://docs.mitmproxy.org/stable/overview-features/ page. I think it's very much on a level with map remote/map local. Sorry if that wasn't clear.
- minor comments inline :)

File: mitmproxy/addons/blocklist.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import mimetypes
+
+from mitmproxy import ctx, exceptions, flowfilter, http, version
+from mitmproxy.net.http.status_codes import RESPONSES
+
+
+class BlockListSpec(typing.NamedTuple):
+    matches: flowfilter.TFilter
+    block_type: str
+    status: str
```
[the remark this answers is not in the document]
Comment (ericbeland) [reply mitmproxy/addons/blocklist.py:11]: That's a good point. I think I tried setting it to an int and perhaps had... some other issue? I'll look again.

File: mitmproxy/addons/blocklist.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            return
+
+        for spec in self.block_list_items:
+            matching = spec.matches(flow)
+            if (matching and spec.block_type == 'block') or (not matching and spec.block_type == 'allow-only'):
+                self.__block_it(flow, spec)
+
+    def __block_it(self, flow, spec):
+        status = int(spec.status)
+        if status == 444:
```
[the remark this answers is not in the document]
Comment (ericbeland) [reply mitmproxy/addons/blocklist.py:53]: Hah. That was done at two different points in time. Will fix.

File: mitmproxy/addons/blocklist.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ericbeland) [reply mitmproxy/addons/blocklist.py:11]: It was in the spec as a string, but I had it converting at both set time and to an int before use. Carrying it as an int is nicer. I realized that the other issue the first time was the typing on the tuple conflicting. Types. Woops!

**Comment by ericbeland** [comment]
Updated per feedback. Re: the allow list, I had been thinking in terms of negating the regexps allowed in filter expressions (which would be ugly), but hadn't stopped to consider whether negating the filter was a thing. This is nicer all-around.

**Comment by ericbeland** [comment]
@mhils  I feel like this is almost ready, but I'd like to add a flag to indicate a request was blocked by the blocklist. I'm not sure where this should go exactly--on flow, probably? The purpose is to allow other addons to identify requests that were stopped by the blocklist, as opposed to those that completed normally--for example, if you are creating a HAR you might not want these in your stats.

**Comment by ericbeland** [comment]
Ok, added it in metadata. This is what I have for now. If there is more feedback, just let me know.

**Pull request merged**